src/core/skills/characters/yijiang2011/xuanhuo.ts

```diff
@@ -1,192 +1,187 @@
-          cardId: response.cardId,
-          triggeredBySkills: [this.Name],
-        };
-
-        await room.useCard(slashUseEvent);
-      } else {
-        await this.doSlashFail(room, fromId, toId);
+    for (const card of cards) {
+      attackCards.push(card.Name);
+    }
+    attackCards.push('duel');
+
+    for (const acard of attackCards) {
+      if (
+        room
+          .getPlayerById(first)
+          .canUseCardTo(
+            room,
+            acard !== 'duel' ? new CardMatcher({ name: [acard] }) : new CardMatcher({ generalName: [acard] }),
+            second,
+          )
+      ) {
+        options.push('xuanhuo:attack');
+        break;
       }
     }
 
-    return true;
-  }
-
-  private async doSlashFail(room: Room, fromId: PlayerId, toId: PlayerId) {
-    const to = room.getPlayerById(toId);
-    const ownedCards = to.getPlayerCards();
+    if (room.getPlayerById(first).getCardIds(PlayerCardsArea.HandArea).length > 0) {
+      options.push('xuanhuo:give');
+    }
 
-    if (ownedCards.length <= 0) {
-      return;
+    if (options.length === 0) {
+      return false;
     }
 
-    const numOfObtain = Math.min(ownedCards.length, 2);
-
-    const cardsChooseEvent: ServerEventFinder<GameEventIdentifiers.AskForChoosingCardEvent> = {
-      amount: numOfObtain,
-      customCardFields: {
-        [PlayerCardsArea.HandArea]: to.getCardIds(PlayerCardsArea.HandArea).length,
-        [PlayerCardsArea.EquipArea]: to.getCardIds(PlayerCardsArea.EquipArea),
-      },
-      toId: fromId,
-      customTitle: TranslationPack.translationJsonPatcher(
-        '{0}: please choose {1} cards to obtain',
+    const askForChooseEvent = EventPacker.createUncancellableEvent<GameEventIdentifiers.AskForChoosingOptionsEvent>({
+      options,
+      conversation: TranslationPack.translationJsonPatcher(
+        '{0}: please choose xuanhuo options: {1}',
         this.Name,
-        numOfObtain,
-      ).toString(),
-    };
+        TranslationPack.patchPlayerInTranslation(room.getPlayerById(second)),
+      ).extract(),
+      toId: first,
+    });
 
     room.notify(
-      GameEventIdentifiers.AskForChoosingCardEvent,
-      EventPacker.createUncancellableEvent<GameEventIdentifiers.AskForChoosingCardEvent>(cardsChooseEvent),
-      fromId,
+      GameEventIdentifiers.AskForChoosingOptionsEvent,
+      EventPacker.createUncancellableEvent<GameEventIdentifiers.AskForChoosingOptionsEvent>(askForChooseEvent),
+      first,
     );
 
-    const { selectedCards, selectedCardsIndex } = await room.onReceivingAsyncResponseFrom(
-      GameEventIdentifiers.AskForChoosingCardEvent,
-      fromId,
-    );
+    const response = await room.onReceivingAsyncResponseFrom(GameEventIdentifiers.AskForChoosingOptionsEvent, first);
+    response.selectedOption = response.selectedOption || askForChooseEvent.options[0];
+
+    if (response.selectedOption === 'xuanhuo:attack') {
+      const newOptions: string[] = [];
+      for (const acard of attackCards) {
```

Comment:
you repeat the operation for the same purpose, you can prepare all the data you need for the first time